File: engine/java/franca/graphics/animations/Tween.java
package java.franca.graphics.animations;

import contracted.java.franca.TranspilableClass;
import java.franca.graphics.device.Router;
import java.franca.graphics.views.Page;
import java.franca.graphics.views.View;

public class Tween extends TranspilableClass {

  Page page = null;
  View view = null;
  Ease ease = null;

  public int tweenId = 0;
  public long registeredTime = 0L;
  public Tween previousTween = null;
  public Tween nextTween = null;

  public Tween(Page page, View view, long duration, int tickerType) {
    this.page = page;
    this.view = view;
    switch (tickerType) {
      case Ease.EASE_LINEAR:
        ease = new LinearEase(0, 100, duration);
        break;
      default:
        // repaint just once
        ease = null;
        break;
    }
  }

  public void setEase(Ease ease) {
    this.ease = ease;
  }

  public Ease getEase() {
    return ease;
  }

  public boolean needsRepainting(Router router) {
    if (ease == null) {
      // one shot animation
      return true;
    }
    return ease.tick(router, registeredTime);
  }

  public boolean needsNextRepainting(Router router) {
    if (ease == null) {
      // one shot animation
      return false;
    }
    return ease.isRunning();
  }
}

File: platforms/typescript/src/java-franca/animations/tween.ts
import {Page} from "../views/page";
import {View} from "../views/view";
import {Ease} from "./ease";
import {LinearEase} from "./linear-ease";
import {Router} from "../device/router";

export class Tween {
  page: Page | null = null;
  view: View | null = null;
  ease: Ease | null = null;

  tweenId: number = 0;
  registeredTime: number = 0;
  previousTween: Tween | null = null;
  nextTween: Tween | null = null;

  constructor(page: Page | null, view: View | null, duration: number, tickerType: number) {
    this.page = page;
    this.view = view;
    switch (tickerType) {
      case Ease.EASE_LINEAR:
        this.ease = new LinearEase(0, 100, duration);
        break;
      default:
        // repaint just once
        this.ease = null;
        break;
    }
  }

  setEase(ease: Ease): void {
    this.ease = ease;
  }

  getEase(): Ease | null {
    return this.ease;
  }

  needsRepainting(router: Router): boolean {
    if (this.ease === null) {
      // one shot animation
      return true;
    }
    return this.ease.tick(router, this.registeredTime);
  }

  needsNextRepainting(router: Router): boolean {
    if (this.ease === null) {
      // one shot animation
      return false;
    }
    return this.ease.isRunning();
  }
}
